Redis Stack use cases (#65)
redis-stack/redis-stack-docs: docs/stack/use-cases/_index.md

> ---

Reviewer: I'm not sure all should be on the same page.
Why not have use-cases under RedisGraph, RediSearch, etc., and here just point to all these pages.

Author: Good idea! I think we need intro paragraph for each section and link to the respective modules. But I also see the value in having them all on the same page.

Author: Let's revisit in the next quarter. I think it may be good to have them on the same page.

- In which store in my city can I locate this item?

**Geo navigation (commerce and travel)**

- List all coffee shops in an area of 2 km, with an average rating higher than 3.5.
- List all parks around me that my friends have been to.

**Operations management (networks and IT)**

- Which parts of the network does this customer rely on?
- Which customers/applications will be affected if this network element fails?
- Is there redundancy in the network for this customer?

**Fraud detection**

- Fraud in health care: Are there any doctors/clinics that are prescribing an unusually high number of a certain drug?

- Fraud in telecom: Does this call/SMS follow a regular calling pattern? It is a new account that just dialed an unusually high number of people, with a short average call-time and was rejected many times, so the receiver of the call should be warned by a real-time notification.

- Fraud in finance: Is this transaction between sender A and recipient B valid? Has the card used by the sender been reported as stolen? Has the device that the recipient is using ever been connected to a fraudulent account?

### Learn more

[RedisGraph](/docs/stack/graph)

## RedisTimeSeries

> 

Reviewer: I think that none the examples below mention how multiple time series can be used.
Multiple airplanes, multiple stocks, etc.

Also, typical queries are not discussed. RedisTimeSeries is not just about storing data, but also about answering questions. What are these questions for each of these use cases?

Author: I agree. Happy to work with you on that.

Reviewer: Would you like to make some suggestions? See my example for environmental measurements above.

### Patterns and use cases

**Monitoring (data center)**

> Modern data centers have a lot of moving pieces, such infrastructure (servers and networks) and software systems (applications and services) that need to be monitored around the clock.

Author: ```suggestion
Modern data centers have a lot of moving pieces, such as infrastructure (servers and networks) and software systems (applications and services) that need to be monitored around the clock.
```

RedisTimeSeries allows you to plan for new resources upfront, optimize the utilization of existing resources, reconstruct the circumstances that led to outages, and identify application performance issues by analyzing and reporting on the following metrics:

- Maximum CPU utilization per server
- Maximum network latency between two services
- Average IO bandwidth utilization of a storage system
- 99th percentile of the response time of a specific application outages

**Weather analysis (environment)**

> A simple use case of using RedisTimeSeries to measure environment is to track how many hours of sunshine and how much rain is observed over the last several years. Seasonally, you can measure average rain, average temperature, maximum of sunny hours, for example. Watch the increase of the maximum temperature over the years. Predict weather in a specific region for a particular week of the year. Example metrics include:

Reviewer: wording: to measure environment

Author: What's your suggestion?

Reviewer: RedisTimeSeries can be used to track environmental measurements such as the number of daily sunshine hours and hourly rainfall depth, over a period of many years. Seasonally, you can measure average rainfall depth, average daily temperature, and the maximum number of sunny hours per day, for example. Watch the increase of the maximum daily temperature over the years. Predict the expected temperature and rainfall depth in a specific location for a particular week of the year. 

Multiple time series can be collect, each for a different location. By utilizing secondary indexes, measurements can be aggregated over given geographical regions (e.g., minimal and maximal daily temperature in Europe) or over locations with specific attributes (e.g., average rainfall depth in mountainous regions).

Example metrics include: ...

**Stock trading (finance)**

Stock trading is highly automated today. Algorithms, and not just human beings, are trading, from the amount of bids and asks for the trading of a stock to the extreme volumes of trades per second (millions of ops per second). Computer-driven trading requires millisecond response times. It's necessary to keep a lot of data points within a very short period of time (for example, price fluctuations per second within a minute). In addition, the long-term history needs to be kept to make statements about trends or for regulatory purposes.  

Use RedisTimeSeries to identify correlations between the trading behavior and other events (for example, social network posts). Discover a developing market. Detect anomalies to discover insider trades. Example metrics include:

- Exact time and order of a trade by itself
- Type of the event (trade/bid)
- The stock price

### Learn more

[RedisTimeSeries](/docs/stack/timeseries)

## RedisBloom

Probabilistic data structures provide a variety of different properties that can have applications across many different industries. 

> ### Bloom filters 

Reviewer: Is there a reason for having separate use cases for Bloom and Cuckoo filters?

Author: None other than that is how they are presented in the technical enablement course for the field. What's your take?

Reviewer: Bloom and Cuckoo filters basically solve the same problem.

In Redis implementation, CF has some additional functionality over BF: [CF.ADDNX](https://redis.io/commands/cf.addnx/), [CF.INSERTNX](https://redis.io/commands/cf.insertnx/), [CF.COUNT](https://redis.io/commands/cf.count/), [CF.DEL](https://redis.io/commands/cf.del/). 

I think we should walk over the examples and check if indeed these are utilized.

There are also some technical differences that are summarized nicely [here](https://brilliant.org/wiki/cuckoo-filter/)

Author: Joined them together!

**Trending hashtags (social media platforms, news distribution networks)** 

This application answers these questions: 

- What are the K hashtags people have mentioned the most in the last X hours? 
- What are the K news with highest read/view count today? 

Data flow is the incoming social media posts from which you parse out the different hashtags. 

The `TOPK.LIST` command has a time complexity of `O(K)` so if `K` is small, there is no need to keep a separate set or sorted set of all the hashtags. You can query directly from the Top K itself. 
 
### HyperLogLogs 
 
**Anonymous unique visits of a web page (SaaS, analytics tools)** 

This application answers these questions: 

- How many unique visits has this page had on this day? 
- How many unique users have played this song? 
- How many unique users have viewed this video? 

<note><b>Note</b> Storing the IP address or any other kind of personal identifier is against the law in some countries, which makes it impossible to get unique visitor statistics on your website.</note> 

One HyperLogLog is created per page (video/song) per period, and every IP/identifier is added to it on every visit.

>  

Reviewer: We need to add t-digest as well.

Author: Sounds good. I'll add that to the list of things we could add.

Author: Added!

- Increased productivity, security, and compliance  

### Learn more

[RediSearch](/docs/stack/search)

## RedisJSON

You can of course use Redis native data structures to store JSON objects, and that's a common practice. For example, you can serialize JSON and save it in a Redis String.

However, RedisJSON provides several benefits over this approach.

### Access and retrieval of subvalues

With RedisJSON, you can get nested values without having to transmit the entire object over the network. Being able to access sub-objects can lead to greater efficiencies when you're storing large JSON objects in Redis.

### Atomic partial updates

RedisJSON allows you to atomically run operations like incrementing a value, adding, or removing elements from an array, append strings, and so on. To do the same with a serialized object, you have to retrieve and then reserialize the entire object, which can be expensive and also lack atomicity.

### Indexing and querying

When you store JSON objects as Redis strings, there's no good way to query those objects. On the other hand, storing these objects as JSON using RedisJSON lets you index and query them. This is provided by RediSearch.

> 

Author: ```suggestion
```

### Application search and external secondary index 

RediSearch supports application search when the application or service uses another primary database. You can use RediSearch as an external secondary index (for example, for indexing numeric or full-text data) by referencing data items within the primary database. Because such secondary indexes are updated asynchronously, you end up with a stale index. However, you can use database synchronization strategies provided by RedisGears.

### Primary database and internal secondary index

RediSearch supports indexing data within Redis. Hashes and RedisJSON documents in Redis represent actual data items. You can declare secondary indexes to support various queries on the data in Redis. RediSearch 2.0 updates indexes automatically whenever a hash/JSON document (that matches the indexes) is updated. 

### Geo-distributed search

In geo-distributed search, hashes/JSON documents are handled in the usual [Active-Active manner](https://docs.redis.com/latest/rs/databases/active-active/). The index follows whatever is written in the documents in the database. Create an index on each database. Then, add synonyms (if used) to each database. 

> <note><b>Note</b> Dictionaries are not supported in Active-Active.</note>

Author: Yeah, I couldn't find any reference to this in the documentation. Sounds out of place here. Going to delete if that's OK.

The application/service creates a temporary and user-specific full-text index in RediSearch when a user logs in. The application/service has direct access to the user-specific index and the primary datastore. When the user logs out of the service, the index is explicitly removed. Otherwise, the index expires after a while (for example, after the user's session expires). 

Using RediSearch for this type of application provides these benefits: 

- Search index is only populated when needed 
- Only a small portion (for example, 2%) of users are actually active 
- Users are only active for a comparatively short period of time
- Small number of documents indexed and so very cost effective in comparison to a persistent search index 

**Real-time inventory (retail)**

In real-time inventory retail, the key question is product availability, "What is available where?" The challenges with such projects are performance and accuracy. RediSearch allows for real-time searching and aggregations over millions of store/SKU combinations.

> Establish real-time event capture from legacy inventory system to RediSearch. Have several inventory services query RediSearch. Use combined queries, for example, item counts, price ranges, categories, and locations. Use geo-distributed search (Active-Active). 

Author: Let's put it through and address in the next quarter.

Using RediSearch for this type of application provides these benefits: 

- Low-latency queries for downstream consumers like marketing, stores/e-commerce, and fulfillment 
- Immediate and higher consistency between stores and datacenters 
- Improved customer experience 
- Real-time/more sensible pricing decisions 
- Less shopping cart abandonment 
- Less remediation (refund, cancellation) 

> **Content recommendations and related products (retail, news, and media)**

Author: Removed!

Gather data using connection information gathering (source IPs, DNS) and conversation data gathering (Wireshark/TShark live capture). Then filter, transform, and store the conversation data in RediSearch to perform search queries and create custom dashboards for your analyses.

Using RediSearch for this type of application provides these benefits: 

- Insights into performance issues, security threats, and network faults 
- Improved service uptime & security 

**Research portal (academia)**

Knowledge workers use research portals to search for articles, research, specs, past solutions, and data to answer specific questions and leverage existing corporate knowledge and history. 

Use unified search and leverages various RediSearch index structures that support tag queries, numeric range queries, geo-location queries, and full-text search (FTS). 

Using RediSearch for this type of application provides these benefits: 

- Provide knowledge workers with only the most relevant, personalized results all while enforcing internal and regulatory data governance policies 
- Increased productivity, security, and compliance  

> **Customer service portal (commerce)**

Author: Done!

Customer support and client service teams use customer service portals to search customer history during support interactions or look for customers matching specific roles or behaviors. Customers also use such portals to search for articles, discussions, and other resources for help and self-serve resolution. 

Using RediSearch for this type of application provides these benefits: 

- Resolve customer issues quickly with immediate access to highly relevant results, thus detecting the creation of new support cases 
- Increased productivity and customer satisfaction 

> **Customer insights (commerce)**

Author: Done!

- One comprehensive view on your customer 
- Allows to incorporate details in order to prioritize or customize the customer experience 
- Less churn and increased customer satisfaction 

### Learn more

[RediSearch](/docs/stack/search)

## RedisJSON

Technically speaking, you can use Redis native data structures to store JSON objects, and that's something users have indeed been doing. It is possible to serialize your JSON object and save it in a simple Redis String, or a Redis Hash, if you have a flat JSON with no nesting. You can even use a combination of the two approaches to store nested JSON in hashes. These are still valid options for storing JSON documents and you need to be able to determine when they are a better fit than using RedisJSON.
These are the two main benefits of RedisJSON over the approaches described above.

### Access and retrieval of subvalues

You can get nested values without having to pull the whole object out of memory, take it to the application layer, deserialize it, and serve the value you need. The overhead of this process is especially prominent for large JSON objects.

### Atomic partial updates

> RedisJSON allows you to atomically run operations like incrementing a value, adding, or removing elements from an array, append strings and so on. To do the same with a serialized object you would have to pull the value out and then write the new value back, which is not atomic.

Reviewer 2: ```suggestion
RedisJSON allows you to atomically run operations like incrementing a value, adding, or removing elements from an array, append strings, and so on. To do the same with a serialized object, you have to retrieve and then reserialize the entire object, which can be expensive and also lack atomicity.

### Indexing and querying

When you store JSON objects as Redis strings, there's no good way to query those objects. On the other hand, storing these objects as JSON using RedisJSON lets you index and query them. This is provided by RediSearch.
```

> On the other hand, you should consider the space complexity/memory overhead of using a more complex data structure such as RedisJSON that splits the document up into a binary tree. In some measurements, RedisJSON can use around 4 to 6 times more memory to represent the same object, compared to a string.

Reviewer 2: ```suggestion

```

Reviewer 2: I think we should remove this caveat for now.

### Learn more

[RedisJSON](/docs/stack/json)

## RedisGraph

> ### Industry-specific examples

Reviewer 2: ```suggestion
### Patters and use cases
```

- In which store in my city can I locate this item?

**Geo navigation (commerce and travel)**

- List all coffee shops in an area of 2 km, with an average rating higher than 3.5.
- List all parks around me that my friends have been to.

**Operations management (networks and IT)**

- Which parts of the network does this customer rely on?
- Which customers/applications will be affected if this network element fails?
- Is there redundancy in the network for this customer?

**Fraud detection**

- Fraud in health care: Are there any doctors/clinics that are prescribing an unusually high number of a certain drug?

- Fraud in telecom: Does this call/SMS follow a regular calling pattern? It is a new account that just dialed an unusually high number of people, with a short average call-time and was rejected many times, so the receiver of the call should be warned by a real-time notification.

- Fraud in finance: Is this transaction between sender A and recipient B valid? Has the card used by the sender been reported as stolen? Has the device that the recipient is using ever been connected to a fraudulent account?

### Learn more

[RedisGraph](/docs/stack/graph)

## RedisTimeSeries

> ### Industry-specific examples 

Reviewer 2: ```suggestion
### Patterns and use cases
```

- Only a small portion (for example, 2%) of users are actually active 
- Users are only active for a comparatively short period of time
- Small number of documents indexed and so very cost effective in comparison to a persistent search index 

**Real-time inventory (retail)**

In real-time inventory retail, the key question is product availability: "What is available where?" The challenges with such projects are performance and accuracy. RediSearch allows for real-time searching and aggregations over millions of store/SKU combinations.

Establish real-time event capture from legacy inventory system to RediSearch. Have several inventory services query RediSearch. Use combined queries, for example, item counts, price ranges, categories, and locations. Use geo-distributed search (Active-Active). 

Using RediSearch for this type of application provides these benefits: 

- Low-latency queries for downstream consumers like marketing, stores/e-commerce, and fulfillment 
- Immediate and higher consistency between stores and datacenters 
- Improved customer experience 
- Real-time/more sensible pricing decisions 
- Less shopping cart abandonment 
- Less remediation (refund, cancellation) 

> **Content recommendations and related products (retail, news, and media)**

Author: ```suggestion
```

> 

Author: ```suggestion
```

> When your goal is to provide content recommendations and related products, you want to present a viewer with related information. The result return must be fast. You need to be able to distinguish co-occurrences of curated topics or keywords (for example, two products are tagged with the keywords "outdoor furniture"). You want to return curated or promoted product combinations. The results must derive from analytics or recommendation models but be cached in a database. And you want to return model-driven associations produced by executing the model on a variety of inputs gathered on demand. You need faceted search algorithms to find top related products. 

Author: ```suggestion

```

> 

Author: ```suggestion
```

> To solve this problem, deep-learning models are used in front of RediSearch for named entity recognition to identify and categorize key information or entities in text. Relevant details are captured as RediSearch hashes/RedisJSON documents. Categories are stored as tags (facets). 

Author: ```suggestion

```

> 

Author: ```suggestion
```

> Using RediSearch for this type of application provides these benefits: 

Author: ```suggestion
```

> 

Author: ```suggestion
```

> - Critical fast subsequent query for the output of a recommendation model 

Author: ```suggestion
```

> - Better user experience and potential sales increase

Author: ```suggestion
```

Gather data using connection information gathering (source IPs, DNS) and conversation data gathering (Wireshark/TShark live capture). Then filter, transform, and store the conversation data in RediSearch to perform search queries and create custom dashboards for your analyses.

Using RediSearch for this type of application provides these benefits: 

- Insights into performance issues, security threats, and network faults 
- Improved service uptime & security 

**Research portal (academia)**

Research portals let users search for articles, research, specs, past solutions, and data to answer specific questions and take advantage of existing knowledge and history. 

To build such a system, you can use RediSearch indexes supporting tag queries, numeric range queries, geo-location queries, and full-text search (FTS). 

Using RediSearch for this type of application provides these benefits: 

- Create relevant, personalized search experiences all while enforcing internal and regulatory data governance policies 
- Increased productivity, security, and compliance  

> **Customer service portal (commerce)**

Author: ```suggestion
```

> 

Author: ```suggestion
```

> Customer support and client service teams use customer service portals to search customer history during support interactions or look for customers matching specific roles or behaviors. Customers also use such portals to search for articles, discussions, and other resources for help and self-serve resolution. 

Author: ```suggestion
```

> 

Author: ```suggestion
```

> Using RediSearch for this type of application provides these benefits: 

Author: ```suggestion
```

> 

Author: ```suggestion
```

> - Resolve customer issues quickly with immediate access to highly relevant results, thus detecting the creation of new support cases 

Author: ```suggestion
```

> - Increased productivity and customer satisfaction 

Author: ```suggestion
```

> 

Author: ```suggestion
```

> **Customer insights (commerce)**

Author: ```suggestion
```

> 

Author: ```suggestion
```

> Customer insights builds on the customer service portal use case. Collate and view all relevant information as it pertains to customer behavior, purchases, and other interactions pulling from a variety of sources systems: CRM, support tickets, database records, documents, social media, and transaction logs. 

Author: ```suggestion
```

> 

Author: ```suggestion
```

> Leverage unified search and application search for this type of application.

Author: ```suggestion
```

> 

Author: ```suggestion
```

> Using RediSearch for this type of application provides these benefits: 

Author: ```suggestion
```

> 

Author: ```suggestion
```

> - One comprehensive view on your customer 

Author: ```suggestion
```

> - Allows to incorporate details in order to prioritize or customize the customer experience 

Author: ```suggestion
```

> - Less churn and increased customer satisfaction 

Author: ```suggestion
```
